### DISCUSSION TITLE
Connection refused

### DISCUSSION THREAD

**Discussion by dszymczuk** [QuorumDMS/ftp-srv#197] [Q&A]
Hi,

I'm trying to run FTP server with docker. I've run on localhost, but I can't run on container on server.
I've opened ports `20-21` and `8400-8500`.
It's dedicated server, so I don't have blocked any ports.

I've got 425 error.

> Status:      	Connecting to MY_EXTERNAL_SERVER_IP:21...
> Status:      	Connection established, waiting for welcome message...
> Status:      	Insecure server, it does not support FTP over TLS.
> Status:      	Logged in
> Status:      	Retrieving directory listing...
> Command:	PWD
> Response: 	257 "/"
> Command:	TYPE I
> Response: 	200 Switch to "binary" transfer mode.
> Command:	PASV
> Response: 	227 PASV OK (MY,EXTERNAL,SERVER,IP,4,3)
> Command:	LIST
> Error:        	The data connection could not be established: ECONNREFUSED - Connection refused by server
> Response: 	425 No connection established
> Error:        	Failed to retrieve directory listing

I can't see any extra logs to check what happened.
Here is ma configuration which I passed via environment variables:

> FTP_HOST=0.0.0.0
> FTP_PORT=21
> FTP_USER=user
> FTP_PASS=pass
> FTP_PASV_URL=MY_EXTERNAL_SERVER_IP
> FTP_PASV_MIN=8400
> FTP_PASV_MAX=8500

Here is my server code:
https://gist.github.com/dszymczuk/37c5ee5754fa006dbb967af78d657cad

**Comment by brucemcpherson** [comment]
I have this running on both kubernetes, and also local docker.
`docker run --env-file env.list -p 18021-18022:18021-18022 -p 18101-18104:18101-18104  gcr.io/<myimage>:td`

and env variables
```
URL=ftp://0.0.0.0:18021
PASVURL=ftp://0.0.0.0
PASVMIN=18101
PASVMAX=18104
RUNMODE=td
```
It all works but after the first 3 files, I get
`ConnectorError: Unable to find valid port`

Running locally, i can send endless amounts of files.

But on Docker (and therefore kubernetes) the ports don't seem to be released after use, even when closing, and exiting the client.

Please let me know if you get this working - I'm stuck.

**Reply by almgwary** [reply]
https://github.com/QuorumDMS/ftp-srv/issues/312  +1

**Comment by brucemcpherson** [comment]
The problem is related to whitelist - I'm minimizing mine to this
```
const whitelist = [
  'STOR',
  'PASS',
  'PASV',
  'USER',
  'TYPE'
]
```
So clearly I'm blocking something that would otherwise free up the port it had just used.
Turns out that the FEAT command needs to be whitelisted to enable the passive port to be released. This is my final working minimum whitelist for uploading files. - works locally, docker and kubernetes
```
const whitelist = [
  'STOR',
  'PASS',
  'PASV',
  'USER',
  'TYPE',
  'QUIT',
  'ABOR',
  'FEAT'
]
```

**Comment by pvtpogopuschel** [comment]
I have experienced similar issues in regards of port starvation when running in docker (exactly like @brucemcpherson). The ports don't get freed after a successful PASV upload and after some time, there are no ports available anymore. However, I am not using the whitelist or blacklist feature. Could this still be the source?

**Comment by brucemcpherson** [comment]
Working code is here if it helps.
https://github.com/brucemcpherson/sensor

**Reply by almgwary** [reply]
Does it work with docker ?

**Comment by matt-forster** [comment]
@pvtpogopuschel We've experienced this in production as well, but haven't found a reason or cause. I _think_ we've avoided it (we haven't run into the problem in 6+ months). I will see what the basic definition of the server looks like and share it, although I think we are using a blacklist. 

We'd definitely welcome help on this front.

**Comment by pvtpogopuschel** [comment]
@brucemcpherson Thanks for the working example! Before reading this bug ticket, my best theory was related to issues arising from my custom file system implementation (and maybe its failure to forward some relevant event to close the connection).  
@forstermatth Are you using a custom file system by any chance as well (or do you have an idea on what you did to circumvent the issue)?   

// Edit: I have considered to Tap into the STOR command to see if this promise ever resolve at some point in time: https://github.com/trs/ftp-srv/blob/master/src/commands/registration/stor.js#L60 and what effect this has on the PASV Connector. 

// Edit2: There also seems to be a old Pull Request related to this issue (which was closed) - https://github.com/trs/ftp-srv/pull/109

**Comment by almgwary** [comment]
@trs

Is there any solution for this as am having same issue after uploading multible files

```
{"name":"ftp-srv","hostname":"77afff52cc2d","directive":"PASV","level":50,"err":{"message":"Unable to find valid port","name":"ConnectorError","stack":"ConnectorError: Unable to find valid port\n    at tryGetPort (/node_modules/ftp-srv/src/helpers/find-port.js:28:16)\n    at Server.<anonymous> (/node_modules/ftp-srv/src/helpers/find-port.js:37:11)\n    at Object.onceWrapper (events.js:421:26)\n    at Server.emit (events.js:314:20)\n    at emitErrorNT (net.js:1343:8)\n    at processTicksAndRejections (internal/process/task_queues.js:84:21)","code":400},"msg":"Unable to find valid port","time":"2022-06-07T14:17:29.308Z","v":0}

```

**Comment by almgwary** [comment]
#312
